hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java

```diff
@@ -43,7 +45,7 @@
   /**
    * Represents the current state of node.
    */
-  private final ConcurrentHashMap<NodeState, Set<UUID>> stateMap;
+  private final ConcurrentHashMap<UUID, NodeStatus> stateMap;
```

Comment:
Do you think it makes sense to have a field inside DatanodeInfo of type NodeStatus, so we can always pass the states around as a pair, or should we add two individual fields to DatanodeInfo - nodeHealth and nodeOperationalState?

Also, one other thing to consider, is nodeStateMap originally kept a list of healthy, stale and dead, so it was possible to quickly return all nodes in that state. However now, we need to iterate over the whole list to find those nodes. One reason for this, is that we have 15 different states now instead of 3. If we move nodeStatus into datanodeInfo, it would be more difficult to optimise this later if needed. However it would simplify things if we simply remove this stateMap.

Reply:
It is better to have `NodeStatus` inside `DatanodeInfo` rather than having two separate fields.

 Yes, `stateMap` helped us to easily get the list/count of nodes in a specific state, but with the current changes it is not straight forward to maintain `state -> list of nodes`. In any case we will be iteration over all the available nodes to get list of nodes in a given state. 
The number of nodes in a cluster should not go beyond 3-4 order of magnitude. We can re-visit and optimize this, if we run into any performance issue.

Reply:
even if you have 15x states, the number of nodes is less. if you have 100 nodes, there are only 1500 states, and if you have 1000 nodes, it is 15000 states. It is still trivial to keep these in memory. Here is the real kicker, just like we decided not to write all cross products for the NodeState static functions, we will end up needing lists of only frequently accessed pattern (in mind that would be (in_service, healthy). All other node queries can be retrieved by iterating the lists as needed.